### PR TITLE
ibft package

### PR DESCRIPTION
**Pull request by rminnich**
Tested by generating an ibft and loading it via linux kernel initramfs.

### CODE REVIEW

File: pkg/acpi/acpi.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+func flags(s ...flag) (uint32, error) {
+	var i, bit uint32
+	for _, f := range s {
+		switch f {
+		case "1":
+			i |= 1 << bit
+		case "0":
+			i &= ^(1 << bit)
+		default:
+			return 0, fmt.Errorf("%s is not a valid value: only 0 or 1 are valied", f)
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]:
```suggestion
			return 0, fmt.Errorf("%s is not a valid value: only 0 or 1 are valid", f)
```

File: pkg/acpi/ibft.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+	Signature  [4]byte  `offset:"0" desc:"iBFT Signature for the iSCSI Boot Firmware Table"`
+	Length     uint32   `offset:"4" desc:"Length in bytes of the entire IBFT, including the signature"`
+	Revision   uint8    `offset:"8" desc:"Revision = 1"`
+	Checksum   uint8    `offset:"9" desc:"Entire table must sum to zero"`
+	OEMID      [6]byte  `offset:"10" desc:"OEM ID. All unused trailing bytes must be zero. [acpi-OEMID]"`
+	OEMTableID [8]byte  `offset:"16" desc:"For the iBFT the Table ID is the Manufacturer‟s Model ID. All unused trailing bytes must be zero."`
+	Reserved   [24]byte `offset:"24" desc:"Reserved"`
+}
+
+var rawIBTFHeader = "IBFT\x00\x08\x00\x001\x00ACPIXXACPISUCK1234567890abcdef01234567"
```
Comment (rjoleary) [inline pkg/acpi/ibft.go]:
```suggestion
var rawIBTFHeader = "IBFT\x00\x08\x00\x001\x00ACPIXXACPISUCKS123456789abcdef01234567"
```

File: pkg/acpi/types.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+// These types are used in emitting binary from a JSON.
+// You can serialize from JSON into structs with these types, and then
+// they can be used in a type switch to serialize out different ways.
+// In one case, the sheap, they serialize into the head and the heap.
+type (
+	// These are serialized as:
+	addr    string // 16 byte IP4 or IP6 addr
+	netaddr string // IP addr as above and 2 byte port
+	flag    string // with other flags in the struct as one u32
```
Comment (rjoleary) [inline pkg/acpi/types.go]: Would it makes more sense for flag to be a bool?

Comment (rminnich) [reply pkg/acpi/types.go]: will json deserialize into a bool?

File: pkg/acpi/ibft.go (hunk 2 above)
Comment (rminnich) [reply pkg/acpi/ibft.go]: no, the ACPISUCK exactly matches a field in ACPI. The length has to be 8.

File: pkg/acpi/types.go (hunk 3 above)
Comment (rminnich) [reply pkg/acpi/types.go]: The whole point of this is that it's something JSON will unmarshall into as strings but I can tell what to do from the type switch.

File: pkg/acpi/acpi.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	for _, v := range val {
+		binary.Write(b, binary.LittleEndian, v)
+		Debug("\t %T %v b is %d bytes", v, v, b.Len())
+	}
+	Debug("w: done: b is %d bytes", b.Len())
+}
+
+// uw writes strings as unsigned words to a bytes.Buffer.
+// Currently it only supports 16, 32, and 64 bit writes.
+func uw(b *bytes.Buffer, s string, bits int) error {
```
Comment (hugelgupf) [inline pkg/acpi/acpi.go]: can you take a look at the stuff here? https://godoc.org/github.com/u-root/u-root/pkg/uio

I would have used a uio.NewLittleEndianBuffer in Marshal, and passed the *uio.Lexer around.

From there, you can do stuff like

```go
buf := uio.NewLittleEndianBuffer(nil)

buf.Write8(u8(...))

return buf.Data()
```

If you need alignment, I'd add it to uio.Lexer.

File: pkg/kexec/memory_linux.go [hunk 5]
```diff
@@ -399,6 +422,38 @@ func (m Memory) FindSpace(sz uint) (start uintptr, err error) {
 	return 0, ErrNotEnoughSpace
 }
 
+// FindSpace returns pointer to the physical memory,
```
Comment (n-canter) [inline pkg/kexec/memory_linux.go]:
```suggestion
// FindACPI returns pointer to the physical memory,
```

File: pkg/kexec/memory_linux.go [hunk 6]
```diff
@@ -422,6 +477,47 @@ func (m *Memory) AddKexecSegment(d []byte) (addr uintptr, err error) {
 	return start, nil
 }
 
+// AddKexecSegment adds d to a new kexec segment starting at base
```
Comment (n-canter) [inline pkg/kexec/memory_linux.go]:
```suggestion
// AddKexecSegmentBase adds d to a new kexec segment starting at base
```

File: pkg/kexec/memory_linux.go [hunk 7]
```diff
@@ -422,6 +477,47 @@ func (m *Memory) AddKexecSegment(d []byte) (addr uintptr, err error) {
...
+func (m *Memory) AddKexecSegment1M(d []byte) (addr uintptr, err error) {
+	start := low1M // someday we may want a fancy allocator. Unlikely.
+	m.addKexecSegment(start, d)
+	return start, nil
+}
+
+// AddKexecSegmentRSDP adds a segment that is at 0x4e for the RSDP
+// We don't use any kind of general "page 0 allocator" because we should
+// almost never do anything in page 0.
+func (m *Memory) AddKexecRSDP(addr uintptr) error {
```
Comment (n-canter) [inline pkg/kexec/memory_linux.go]:
```suggestion
func (m *Memory) AddKexecSegmentRSDP(addr uintptr) error {
```

File: pkg/acpi/tables_linux.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+// or a standin for everything.
+
+// Table contains a []byte for a single ACPI table.
+// The Name is provided for convenience.
+type RawTable struct {
+	Name string
+	Data []byte
+}
+
+// Tables returns an array of Table
```
Comment (n-canter) [inline pkg/acpi/tables_linux.go]:
```suggestion
// RawTables returns an array of Table.
```

File: pkg/acpi/rsdp.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+func num(n string, i int) (uint64, error) {
+	b, err := ioutil.ReadFile(fmt.Sprintf("/sys/firmware/memmap/%d/%s", i, n))
+	if err != nil {
+		return 0, err
+	}
+	start, err := strconv.ParseUint(string(b), 0, 64)
+	return start, err
+}
+
+// get RSDPmem is the option of last choice, it just grovels through
```
Comment (n-canter) [inline pkg/acpi/rsdp.go]:
```suggestion
// getRSDPmem is the option of last choice, it just grovels through.
```

File: pkg/acpi/rsdp.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+// Base returns the RSDP base address as an int64
+// We use int64 for compatibility with Go notions
+// of offsets.
+func (r *RSDP) Base() int64 {
+	return int64(r.base)
+}
+
+// Data returns the RSDP as a []byte
```
Comment (n-canter) [inline pkg/acpi/rsdp.go]:
```suggestion
// Data returns the RSDP as a []byte.
```

File: pkg/acpi/rsdp.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+func (r *RSDP) Data() []byte {
+	return r.data[:]
+}
+
+// Sig returns the RSDP signature
```
Comment (n-canter) [inline pkg/acpi/rsdp.go]:
```suggestion
// Sig returns the RSDP signature.
```

File: pkg/acpi/rsdp.go [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+func (r *RSDP) Sig() string {
+	return string(r.data[:8])
+}
+
+// OEMID returns the RSDP OEMID
```
Comment (n-canter) [inline pkg/acpi/rsdp.go]:
```suggestion
// OEMID returns the RSDP OEMID.
```

File: pkg/acpi/rsdp.go [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+func (r *RSDP) OEMID() string {
+	return string(r.data[9:15])
+}
+
+// Revision returns the RSDP revision, which
+// after 2002 should be >= 2
```
Comment (n-canter) [inline pkg/acpi/rsdp.go]:
```suggestion
// after 2002 should be >= 2.
```

File: pkg/acpi/tables_linux.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+		b, err := ioutil.ReadFile(t)
+		if err != nil {
+			return nil, err
+		}
+		tabs = append(tabs, &RawTable{Name: t, Data: b})
+	}
+	return tabs, nil
+}
+
+// TablesData returns all ACPI tables as a single []byte
```
Comment (n-canter) [inline pkg/acpi/tables_linux.go]:
```suggestion
// TablesData returns all ACPI tables as a single []byte.
```

File: pkg/kexec/memory_linux.go [hunk 15]
```diff
@@ -399,6 +422,38 @@ func (m Memory) FindSpace(sz uint) (start uintptr, err error) {
 	return 0, ErrNotEnoughSpace
 }
 
+// FindSpace returns pointer to the physical memory,
+// where array of size sz can be stored during next
+// AddKexecSegment call.
```
Comment (n-canter) [inline pkg/kexec/memory_linux.go]:
```suggestion
// AddKexecSegmentACPI call.
```

**Comment by rminnich** [comment]
this one is dead to me now.

**Pull request closed without merging**